sdk/v2/core/workers.go

```diff
@@ -0,0 +1,381 @@
+	return []WorkerPhase{
+		WorkerPhaseAborted,
+		WorkerPhaseCanceled,
+		WorkerPhaseFailed,
+		WorkerPhaseSucceeded,
+		WorkerPhaseTimedOut,
+	}
+}
+
+// WorkerPhasesNonTerminal returns a slice of WorkerPhases containing ALL phases
+// that are considered non-terminal. Note that instead of utilizing a
+// package-level slice, this a function returns ad-hoc copies of the slice in
+// order to preclude the possibility of this important collection being modified
+// at runtime by a client.
+func WorkerPhasesNonTerminal() []WorkerPhase {
+	return []WorkerPhase{
+		WorkerPhasePending,
+		WorkerPhaseRunning,
+		WorkerPhaseUnknown,
+	}
+}
+
+// Worker represents a component that orchestrates handling of a single Event.
+type Worker struct {
+	// Spec is the technical blueprint for the Worker.
+	Spec WorkerSpec `json:"spec"`
+	// Status contains details of the Worker's current state.
+	Status WorkerStatus `json:"status"`
+	// Jobs contains details of all Jobs spawned by the Worker during handling of
+	// the Event.
+	Jobs map[string]Job `json:"jobs,omitempty"`
+}
+
+// WorkerSpec is the technical blueprint for a Worker.
+type WorkerSpec struct {
+	// Container specifies the details of an OCI container that forms the
+	// cornerstone of the Worker.
+	Container    *ContainerSpec `json:"container,omitempty"`
+	UseWorkspace bool           `json:"useWorkspace"`
+	// WorkspaceSize specifies the size of a volume that will be provisioned as
+	// a shared workspace for the Worker and any Jobs it spawns.
+	// The value can be expressed in bytes (as a plain integer) or as a
+	// fixed-point integer using one of these suffixes: E, P, T, G, M, K.
+	// Power-of-two equivalents may also be used: Ei, Pi, Ti, Gi, Mi, Ki.
+	WorkspaceSize string `json:"workspaceSize,omitempty"`
+	// Git contains git-specific Worker details.
+	Git *WorkerGitConfig `json:"git,omitempty"`
+	// Kubernetes contains Kubernetes-specific Worker details.
+	Kubernetes *WorkerKubernetesConfig `json:"kubernetes,omitempty"`
+	// JobPolicies specifies policies for any Jobs spawned by the Worker.
+	JobPolicies *JobPolicies `json:"jobPolicies,omitempty"`
+	// LogLevel specifies the desired granularity of Worker log output.
+	LogLevel LogLevel `json:"logLevel,omitempty"`
+	// ConfigFilesDirectory specifies a directory within the Worker's workspace
+	// where any relevant configuration files (e.g. brigade.json, brigade.js,
+	// etc.) can be located.
+	ConfigFilesDirectory string `json:"configFilesDirectory,omitempty"`
+	// DefaultConfigFiles is a map of configuration file names to configuration
+	// file content. This is useful for Workers that do not integrate with any
+	// source control system and would like to embed configuration (e.g.
+	// brigade.json) or scripts (e.g. brigade.js) directly within the WorkerSpec.
+	DefaultConfigFiles map[string]string `json:"defaultConfigFiles,omitempty"`
+}
+
+// WorkerGitConfig represents git-specific Worker details.
+type WorkerGitConfig struct {
```

Comment:
Two questions related the worker git object:

- is there a need for it to be conceptually different from a job's git information?
- is `initSubmodules` a property that should only be available for a worker, or also for a job?

Reply:
Related to [this comment](https://github.com/brigadecore/brigade/pull/1118#discussion_r492276970), I'm going to double-check.

Reply:
_At the very least_, it seems the docs need to differentiate more clearly between similarly named things.

Reply:
@radu-matei, so...

`WorkerGitConfig` is used when defining the (`WorkerSpec`) "template" that a Project will use for creating new Workers in response to an Event. This says where to get source (possibly including things like `brigade.js`) from and, yes, indicates whether or not submodules should be initialized. (These are details that the interactive `brig project create` asks about for VCS projects in v1.x.)

What I think is definitely _not_ clear from the name of this type (or from the docs) is that this configuration is applied _not only_ to the init container of a Worker in need of source, but applies _equally_ to init containers of _Jobs_ (spawned by that Worker) if they too are in need of source. (And this, isn't different from 1.x either.)

`EventGitConfig` has a related, but different purpose. It gives individual Events the possibility of overriding the Project-level git config (described above). The `CloneURL`, for instance, may be overridden if the Event should prompt Brigade to clone source from a fork. `Commit` or `Ref` may similarly override Project-level fields of the same name, directing a specific commit, branch, tag, etc. to be pulled. (Again, I think no different from 1.x.) I cannot imagine the matter of whether submodules should or shouldn't be initialized to be something that the Event has any business overriding-- that seems more appropriately determined only by whoever set up the Project.

So there's a relationship between these two things, for sure. And for sure, one of them rightfully has one field that the other lacks and the rest of the fields are common... But it's not clear to me if there's an opportunity to DRY things up through composition here-- _at least not in a way that makes sense (to me)._

We _could_ do this, for instance:

```
type EventGitConfig struct {
  CloneURL string `json:"cloneURL,omitempty"`
  Commit string `json:"commit,omitempty"`
  Ref string `json:"ref,omitempty"`
}

type WorkerGitConfig struct {
  EventGitConfig `json:",inline"`
  InitSubmodules bool `json:"initSubmodules"`
}
```

This is DRYer, but it makes no sense that any Project-level configuration (the Worker template) would be composed of seemingly Event-specific details.

Then I thought about toying with the naming to make it make more sense:

```
type GitConfig struct {
  CloneURL string `json:"cloneURL,omitempty"`
  Commit string `json:"commit,omitempty"`
  Ref string `json:"ref,omitempty"`
}

type WorkerGitConfig struct {
  GitConfig `json:",inline"`
  InitSubmodules bool `json:"initSubmodules"`
}
```

This looks considerably better, I think, until you try to build out a Project and `GitConfig` "stutters" like so:

```
project := Project {
  Spec: ProjectSpec {
    WorkerTemplate: WorkerSpec {
      Git: &WorkerGitConfig {
        GitConfig: GitConfig { ... }, // This stutters, without any intuitive reason for doing so
      },
    },
  },
}
```

So there's the thought process. My conclusion was that clear was better than clever here, even if it was less DRY.

But I am totally open to suggestions for untangling this little knot because even if I think it's the least awkward of the possibilities I considered, I still think it could be improved.

Reply:
Slept on it and had a revelation...

Events don't have configuration; they have _details_.

I believe everything might be more clear if `WorkerSpec` has a field of type `GitConfig` whilst `Event` has a field of type `GitDetails`.

`GitConfig` and `GitDetails` may have fields in common (that repeat), but perhaps that's the small price we pay for favoring clarity over cleverness.

Would love thoughts on this.

Reply:
I am on board with the distinction between `GitConfig` and `GitDetails` you just made.

Reply:
> although it is definitely not blocking this PR.

I'll fix it now. I think it improves clarity substantially.

Reply:
There was an opportunity to fix some similar confusion pertaining to `KubernetesConfig` (and other similarly named types). I fixed that as well.